swarm/src/behaviour.rs

```diff
@@ -248,177 +233,13 @@ pub enum NetworkBehaviourAction<
-    ///         if let FromSwarm::DialFailure(DialFailure { handler, .. }) = event {
-    ///             let msg = handler.message.unwrap();
-    ///             self.outbox_to_swarm
-    ///               .push_back(NetworkBehaviourAction::GenerateEvent(msg))
-    ///         }
-    ///     }
-    ///     #
-    ///     # fn poll(
-    ///     #     &mut self,
-    ///     #     _: &mut Context<'_>,
-    ///     #     _: &mut impl PollParameters,
-    ///     # ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ConnectionHandler>> {
-    ///     #     if let Some(action) = self.outbox_to_swarm.pop_front() {
-    ///     #         return Poll::Ready(action);
-    ///     #     }
-    ///     #     Poll::Pending
-    ///     # }
-    /// }
-    ///
-    /// # struct MyHandler {
-    /// #     message: Option<PreciousMessage>,
-    /// # }
-    /// #
-    /// # impl ConnectionHandler for MyHandler {
-    /// #     type InEvent = Void;
-    /// #     type OutEvent = Void;
-    /// #     type Error = Void;
-    /// #     type InboundProtocol = DeniedUpgrade;
-    /// #     type OutboundProtocol = DeniedUpgrade;
-    /// #     type InboundOpenInfo = ();
-    /// #     type OutboundOpenInfo = Void;
-    /// #
-    /// #     fn listen_protocol(
-    /// #         &self,
-    /// #     ) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
-    /// #         SubstreamProtocol::new(DeniedUpgrade, ())
-    /// #     }
-    /// #
-    /// #     fn on_behaviour_event(&mut self, _event: Self::InEvent) {}
-    /// #
-    /// #     fn on_connection_event(
-    /// #        &mut self,
-    /// #        event: ConnectionEvent<
-    /// #            Self::InboundProtocol,
-    /// #            Self::OutboundProtocol,
-    /// #            Self::InboundOpenInfo,
-    /// #            Self::OutboundOpenInfo,
-    /// #        >,
-    /// #     ) {}
-    /// #
-    /// #     fn connection_keep_alive(&self) -> KeepAlive {
-    /// #         KeepAlive::Yes
-    /// #     }
-    /// #
-    /// #     fn poll(
-    /// #         &mut self,
-    /// #         _: &mut Context<'_>,
-    /// #     ) -> Poll<
-    /// #         ConnectionHandlerEvent<
-    /// #             Self::OutboundProtocol,
-    /// #             Self::OutboundOpenInfo,
-    /// #             Self::OutEvent,
-    /// #             Self::Error,
-    /// #         >,
-    /// #     > {
-    /// #         todo!("If `Self::message.is_some()` send the message to the remote.")
-    /// #     }
-    /// # }
-    /// # #[derive(Debug, PartialEq, Eq)]
-    /// # struct PreciousMessage(String);
-    /// ```
-    Dial { opts: DialOpts, handler: THandler },
+    /// The provided [`ConnectionId`] will be used throughout the connection's lifecycle to associate
+    /// events with it. This allows a [`NetworkBehaviour`] to identify a connection that resulted out
+    /// of its own dial request.
+    Dial {
+        opts: DialOpts,
+        connection_id: ConnectionId,
```

Comment:
Why not make `connection_id` part of `DialOpts`? For those that don't need the `connection_id`, they don't need to bother creating one, as the `DialOpts` builder would create one for them.

Do you favor being explicit, i.e. requiring the user to create the `ConnectionId`, thus being aware of it?

Reply:
We could move it into `DialOpts` I guess, that would make things slightly more ergonomic. Together with https://github.com/libp2p/rust-libp2p/pull/3221, we could then make the constructor `pub(crate)` and thus prevent that users assign existing connection IDs to new connections.

Reply:
Addressed. I also created https://github.com/libp2p/rust-libp2p/pull/3335 which I temporarily merged into this PR to keep going here.